### DISCUSSION TITLE
Action: Can't Response JSON on JSON Post Request

### DISCUSSION THREAD

**Discussion by kicut** [remix-run/remix#4491] [Q&A] [answered]
Hello everyone, I want to send a JSON POST request on the client, I have tried to return JSON in `ActionFunction` via `json()` function, but it seems this function is not returning it correctly. How can I get a JSON response on a POST request?

Stackblitz: https://stackblitz.com/edit/node-mexycj?file=app%2Froutes%2Findex.tsx

```typescript
import { ActionFunction, json } from '@remix-run/node';

export const action: ActionFunction = async ({ request, params }) => {
  if (
    request.method === 'POST' &&
    request.headers.get('content-type') === 'application/json'
  ) {
    return json({ message: 'hello' });
  }
};

export default function Index() {
  const postJson = async () => {
    try {
      const res = await fetch('?index', {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
        },
        body: JSON.stringify({ data: 'test' }),
      });

      const resJson = await res.json();
      console.log('result=====>', resJson);
    } catch (error) {
      // 🚨 error=====> SyntaxError: Unexpected token '<', "<!DOCTYPE "... is not valid JSON
      console.log('error=====>', error);
    }
  };

  return <button onClick={postJson}>Send Post</button>;
}

```

**Comment by sergiodxa** [comment, marked as answer]
To send a request only to a loader or action on a UI route you need to add the `_data` search param on the URL with the route ID, that's why you get an HTML back

The route ID is the path of the route inside the `app` folder without the extension, so in your example is `routes/index`.

I recommend you to use the useFetcher hook instead of doing the fetch yourself.

And instead of using useFetcher I recommend you more to use the Form component, you can wrap your button in the Form component even if it's a single button.

If you still want to do the fetch manually, move the action to a resource route (a route without an `export default`, with a resource route you don't need the `_data` search param anymore. And don't forget to track the request status and avoid race conditions.

**Reply by kicut** [reply]
Thanks @sergiodxa , adding `_data` to the search parameters worked for me.

I have tried using `fethcer.submit`, but since I have to call the function in `useEffect`, it sends the POST request twice, I can't find a way to stop the request using the function like using AbortController, if I use fetch I can stop it using AbortController. I've got a solution to this problem in a [previous question](https://github.com/remix-run/remix/discussions/4469), and I'm looking for the best one.

**Reply by sergiodxa** [reply]
The double request is because you’re most likely using React.StrictMode which makes all effects run twice

Ideally mutations should happen as the result of an user event and not an side effect of rendering.
